components/playlist/playlist_media_file_downloader.cc

```diff
@@ -0,0 +1,354 @@
+
+  const base::FilePath file_path =
+      playlist_dir_path_.Append(source_media_files_dir_)
+          .Append(GetFileNameStringFromIndex(index));
+  iter->get()->DownloadToFile(
+      url_loader_factory_.get(),
+      base::BindOnce(&PlaylistMediaFileDownloader::OnMediaFileDownloaded,
+                     base::Unretained(this), std::move(iter), index),
+      file_path);
+}
+
+void PlaylistMediaFileDownloader::OnMediaFileDownloaded(
+    SimpleURLLoaderList::iterator iter,
+    int index,
+    base::FilePath path) {
+  url_loaders_.erase(iter);
+  VLOG(2) << __func__ << ": downloaded media file at " << path;
+
+  if (path.empty()) {
+    // This fail is handled during the generation.
+    // See |has_skipped_source_files| in DoGenerateSingleMediaFile().
+    // |has_skipped_source_files| will be set to true.
+    VLOG(1) << __func__ << ": failed to download media file at " << index;
+  }
+
+  remained_download_files_--;
+
+  // If all source files are downloaded, unify them into one media file.
+  if (IsDownloadFinished())
+    StartSingleMediaFileGeneration();
+}
+
+void PlaylistMediaFileDownloader::RequestCancelCurrentPlaylistGeneration() {
+  ResetDownloadStatus();
+}
+
+void PlaylistMediaFileDownloader::StartSingleMediaFileGeneration() {
+  base::PostTaskAndReplyWithResult(
+      task_runner(), FROM_HERE,
+      base::BindOnce(&DoGenerateSingleMediaFile, playlist_dir_path_,
+                     source_media_files_dir_, unified_media_file_name_,
+                     media_file_source_files_count_),
+      base::BindOnce(&PlaylistMediaFileDownloader::OnSingleMediaFileGenerated,
+                     weak_factory_.GetWeakPtr(),
+                     current_playlist_id_));
+}
+
+// If some of source files are not fetched properly, it should be treated as
+// fail.
+void PlaylistMediaFileDownloader::OnSingleMediaFileGenerated(
+    const std::string& id, bool success) {
+  // If canceled or canceled and new download is started, |id| will be different
+  // with |current_playlist_id_|.
+  // Just silently end here.
+  if (id != current_playlist_id_)
+    return;
+
+  if (success) {
+    base::FilePath media_file_path =
+        playlist_dir_path_.Append(unified_media_file_name_);
+    NotifySucceed(id, media_file_path_key_,
+                  ConvertFilePathToUTF8(media_file_path));
+  } else {
+    NotifyFail(id);
+  }
+}
+
+base::SequencedTaskRunner* PlaylistMediaFileDownloader::task_runner() {
+  if (!task_runner_) {
+    task_runner_ = base::CreateSequencedTaskRunner(
+        {base::ThreadPool(), base::MayBlock(), base::TaskPriority::BEST_EFFORT,
+         base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
+  }
+  return task_runner_.get();
+}
+
+void PlaylistMediaFileDownloader::ResetDownloadStatus() {
+  in_progress_ = false;
+  remained_download_files_ = 0;
+  media_file_source_files_count_ = 0;
+  current_playlist_id_.clear();
+  current_playlist_ = base::Value();
+  url_loaders_.clear();
```

Comment:
should we also clear `playlist_dir_path_`?

Reply:
Cleared.